contracts/extensions/executor/Executor.sol

```diff
@@ -0,0 +1,114 @@
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+ */
+
+contract ExecutorExtension is DaoConstants, AdapterGuard, IExtension {
+    using Address for address payable;
+
+    bool public initialized = false; // internally tracks deployment under eip-1167 proxy pattern
+    DaoRegistry public dao;
+
+    enum AclFlag {EXECUTE}
+
+    /// @notice Clonable contract must have an empty constructor
+    // constructor() {
+    // }
+
+    modifier hasExtensionAccess(AclFlag flag) {
+        require(
+            address(this) == msg.sender ||
+                address(dao) == msg.sender ||
+                dao.state() == DaoRegistry.DaoState.CREATION ||
+                dao.hasAdapterAccessToExtension(
+                    msg.sender,
+                    address(this),
+                    uint8(flag)
+                ),
+            "executor::accessDenied"
+        );
+        _;
+    }
+
+    /**
+     * @notice Initialises the Executor extension to be associated with a DAO
+     * @dev Can only be called once
+     * @param creator The DAO's creator, who will be an initial member
+     */
+    function initialize(DaoRegistry _dao, address creator) external override {
+        require(!initialized, "executor::already initialized");
+        require(_dao.isMember(creator), "executor::not member");
+        dao = _dao;
+        initialized = true;
+    }
+
+    /**
+     * @notice Delegates the function call to the implementation address provided as function argument.
+     * @dev reference contract @openzeppelin/contracts/proxy/Proxy.sol
+     * @param implementation The address of the contract that will execute the delegated function call.
+     */
+    function delegate(address implementation)
+        external
+        payable
+        hasExtensionAccess(AclFlag.EXECUTE)
+    {
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            // Copy msg.data. We take full control of memory in this inline assembly
+            // block because it will not return to Solidity code. We overwrite the
+            // Solidity scratch pad at memory position 0.
+            calldatacopy(0, 0, calldatasize())
+
+            // Call the implementation.
+            // out and outsize are 0 because we don't know the size yet.
+            let result := delegatecall(
+                gas(),
+                implementation,
+                0,
+                calldatasize(),
+                0,
+                0
+            )
+
+            // Copy the returned data.
+            returndatacopy(0, 0, returndatasize())
+
+            switch result
```

Comment:
if we return some data we should declare it in the function as well.

Looks like you return the same whether it's an error or not. Maybe something is wrong there (not 100% sure)

The approach is the right one. 

I think to write a proper test we should create a dummy adapter that uses the executor. Here is the use case I propose

Create a simple ERC20 with one extra function "mint" that anybody can call (it's for test)
This function mints erc20 for msg.sender

The adapter should do the following:
- mint  10'000 tokens
- send the tokens to the bank address
- from another transaction, it should collect the tokens 
to collect the token, just use the BankAdapter.updateToken

Let me know if you have any questions